### PR TITLE
Initial cut at tower_token module

### PR DESCRIPTION
**Pull request by john-westcott-iv**
##### SUMMARY

Initial creation of tower_token modules. 

I decided to mimic the ovirt module behavior of setting an ansible fact which can then be used directly in the tower_oatuhtoken parameter of other modules.

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

 - Collection

##### AWX VERSION

```
awx: 11.2.0
```

##### ADDITIONAL INFORMATION

```
Blocked by https://github.com/ansible/awx/pull/7118. Once that is merged in we can fix the module linting problems with ease.
```

### CODE REVIEW

File: awx_collection/plugins/modules/tower_token.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+'''
+
+EXAMPLES = '''
+- name: Create a new token using an existing token
+  tower_token:
+    description: '{{ token_description }}'
+    scope: "write"
+    state: present
+    tower_oauthtoken: "{{ ny_existing_token }}"
+  register: new_token
```
Comment (kedark3) [inline awx_collection/plugins/modules/tower_token.py]: Is there a way to create token with username/password ? :) Absolutely Great work otherwise. Sincere thanks for Tower Collections team for taking customer feedback seriously :100:

Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_token.py]: Yes, just use the tower_username/tower_password parameters instead of the tower_oauthtoken.

Comment (kedark3) [reply awx_collection/plugins/modules/tower_token.py]: Awesome. Do you mind adding and example,it will be tremendously useful.

File: awx_collection/plugins/modules/tower_token.py [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+    tower_oauthtoken:
+      description:
+        - The Tower OAuth token to use.
+      required: False
+      type: str
+      version_added: "3.7"
+extends_documentation_fragment: awx.awx.auth
+'''
+
+EXAMPLES = '''
```
Comment (mshriver) [inline awx_collection/plugins/modules/tower_token.py:67]: Thanks for adding the Examples block, good timing.

File: awx_collection/plugins/modules/tower_token.py (hunk 1 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_token.py]: Done.

**Review by beeankha — APPROVED** [review]

File: awx_collection/plugins/doc_fragments/auth.py [hunk 3]
```diff
@@ -32,7 +32,7 @@ class ModuleDocFragment(object):
     description:
     - The Tower OAuth token to use.
     - If value not set, will try environment variable C(TOWER_OAUTH_TOKEN) and then config files
-    type: str
+    type: raw
```
Comment (AlanCoding) [inline awx_collection/plugins/doc_fragments/auth.py:38]: if this type is raw, I want to see some information in the `description:` with direction to the user. When is it not a string, and what format should be expected in that case?

File: awx_collection/plugins/modules/tower_token.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+- block:
+    - name: Create a new token using an existing token
+      tower_token:
+        description: '{{ token_description }}'
+        scope: "write"
+        state: present
+        tower_oauthtoken: "{{ ny_existing_token }}"
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_token.py]: `ny_existing_token` -> `my_existing_token`?

File: awx_collection/plugins/modules/tower_token.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    - name: Use our new token to make another call
+      tower_job_list:
+        tower_oauthtoken: "{{ tower_token }}"
+
+  always:
+    - name: Delete our Token with the token we created
+      tower_token:
+        existing_token: "{{ tower_token }}"
+        state: absent
+      when: tower_token is defined
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_token.py:98]: this looks like a duplicated task with "Delete this token"

File: awx_collection/plugins/modules/tower_token.py [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+      tower_token:
+        description: '{{ token_description }}'
+        scope: "write"
+        state: present
+        tower_username: "{{ ny_username }}"
+        tower_password: "{{ ny_password }}"
+
+    - name: Use our new token to make another call
+      tower_job_list:
+        tower_oauthtoken: "{{ tower_token }}"
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_token.py:91]: I think the first task needs to add `register: result`, and then this task reference `"{{ result.tower_token }}"`, I don't think things can just get crammed into the globals. Not as a module plugin... maybe as an action plugin, but that would be hard to follow.

File: awx_collection/plugins/module_utils/tower_api.py [hunk 7]
```diff
@@ -84,7 +84,10 @@ def __init__(self, argument_spec, **kwargs):
         if self.params.get('validate_certs') is not None:
             self.verify_ssl = self.params.get('validate_certs')
         if self.params.get('tower_oauthtoken'):
-            self.oauth_token = self.params.get('tower_oauthtoken')
+            if type(self.params.get('tower_oauthtoken')) is dict:
+                self.oauth_token = self.params.get('tower_oauthtoken')['token']
+            elif 'token' in self.params.get('tower_oauthtoken'):
+                self.oauth_token = self.params.get('tower_oauthtoken')
```
Comment (AlanCoding) [inline awx_collection/plugins/module_utils/tower_api.py:108]: After this line, the code should validate type of `self.oauth_token`, because it's no longer automatic. Here, only a string type should be acceptable.

File: awx_collection/tests/integration/targets/tower_token/tasks/main.yml [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+    - name: Create a Token
+      tower_token:
+        description: '{{ token_description }}'
+        scope: "write"
+        state: present
+      register: new_token
+
+    - name: Validate our token works by token
+      tower_job_list:
+        tower_oauthtoken: "{{ tower_token.token }}"
```
Comment (AlanCoding) [inline awx_collection/tests/integration/targets/tower_token/tasks/main.yml:66]: I think this should be what was registered, and the documented return variable, `new_token.tower_token`

File: awx_collection/plugins/doc_fragments/auth.py (hunk 3 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/doc_fragments/auth.py:38]: Changed it to:
- tower_oauthtoken
        The Tower OAuth token to use.
        This value can be in one of two formats.
        A string which is the token itself. (i.e. bqV5txm97wqJqtkxlMkhQz0pKhRMMX)
        A dictionary structure as returned by the tower_token module.
        If value not set, will try environment variable `TOWER_OAUTH_TOKEN' and then config files

File: awx_collection/plugins/modules/tower_token.py (hunk 4 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_token.py]: Fixed, there was also a ny_password and ny_username.

File: awx_collection/plugins/modules/tower_token.py (hunk 5 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_token.py:98]: In the example we actually pull two tokens:
* block
  * pull a token
  * delete the token we pulled
  * pull a second token
  * use the second token to call tower_job_list
* always
  * delete the second token

File: awx_collection/plugins/modules/tower_token.py (hunk 6 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_token.py:91]: Up on line 107 we are doing the injection of the tower_token variable into the ansible_facts with this code snippet:
```
    module.json_output['ansible_facts'] = {
        'tower_token': last_response,
    }
```
This is similar the same as to what the ovirt_auth module does: https://github.com/ansible/ansible/blob/stable-2.9/lib/ansible/modules/cloud/ovirt/ovirt_auth.py#L279

File: awx_collection/tests/integration/targets/tower_token/tasks/main.yml (hunk 8 above)
Comment (john-westcott-iv) [reply awx_collection/tests/integration/targets/tower_token/tasks/main.yml:66]: We do document the return token as a dict (but not whats inside of the dict):
```
tower_token:
  type: dict
  description: A Tower token object which can be used for auth or token deletion
  returned: on successful create
```

File: awx_collection/plugins/module_utils/tower_api.py (hunk 7 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/module_utils/tower_api.py:108]: Not sure how you got the code like this, maybe an old commit? It looks like this:
```
        if self.params.get('tower_oauthtoken'):
            if type(self.params.get('tower_oauthtoken')) is dict:
                self.oauth_token = self.params.get('tower_oauthtoken')['token']
            elif 'token' in self.params.get('tower_oauthtoken'):
                self.oauth_token = self.params.get('tower_oauthtoken')
```

Comment (john-westcott-iv) [reply awx_collection/plugins/module_utils/tower_api.py:108]: We could I guess try and raise an exception if we fall down into an "else" condition here. LMK if you want me to do that.

File: awx_collection/plugins/doc_fragments/auth.py (hunk 3 above)
Comment (AlanCoding) [reply awx_collection/plugins/doc_fragments/auth.py:38]: 👍  to that description

but for whatever reason, I'm not seeing that in the current diff

File: awx_collection/plugins/module_utils/tower_api.py (hunk 7 above)
Comment (AlanCoding) [reply awx_collection/plugins/module_utils/tower_api.py:108]: yeah, I'm also not really quite sure what's going on here, and if you actually intended this diff.

Backing up, what you're doing is turning off the validation for the token parameter. That's fine is documented, and we give clean error messages when it doesn't fit the expected pattern.

In several ways, it seems like this still isn't finished, or at least the validation isn't.

What are our expectations?

 - If it is a dict, then it needs to have the key "token", it can have other keys, but this is a minimum condition. Ansible handles `KeyError`s poorly, so if it's a dict, but that dict doesn't have the token key, this should fail with a message to document the failure of expectations.
 - If it's a string, then it should be used as-is
 - If it's a list, or any other type, then it should fail, repeating or referring to the basic documentation of the token parameter

The current code has a few head-scratchers. where is it handed as a string? `elif 'token' in self.params.get('tower_oauthtoken'):` is assuming a dict, but the prior conditional is also assuming a dict.

File: awx_collection/plugins/doc_fragments/auth.py (hunk 3 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/doc_fragments/auth.py:38]: @beeankha asked me to hold off on the push until she finished up a yolo run. This will be up later today.

File: awx_collection/plugins/modules/tower_token.py [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+      tower_token:
+        existing_token: "{{ tower_token }}"
+        state: absent
+      when: tower_token is defined
+'''
+
+RETURNS = '''
+tower_token:
+  type: dict
+  description: A Tower token object which can be used for auth or token deletion
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_token.py]: The ovirt example you link documents the type as `complex` and gives the sub-keys

https://github.com/ansible/ansible/blob/stable-2.9/lib/ansible/modules/cloud/ovirt/ovirt_auth.py#L158

I actually think that would be good. Even if it gave more keys than what are documented, this would be a good way to specify the 2 functional items - the token and the token_id. In the parameter documentation we discussed

> A dictionary structure as returned by the tower_token module.

It just defers here. We also don't have any real examples of the dictionary typed out. I think it needs to be somewhere more clearly stated that (1) the function of this data structure is to contain the token and the token id, and what the key names for those are and (2) also mention in the return docs that this is returned both in this key and in the ansible_facts.

File: awx_collection/plugins/modules/tower_token.py [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+
+DOCUMENTATION = '''
+---
+module: tower_token
+author: "John Westcott IV (@john-westcott-iv)"
+version_added: "2.3"
+short_description: create, update, or destroy Ansible Tower tokens.
+description:
+    - Create or destroy Ansible Tower tokens. See
+      U(https://www.ansible.com/tower) for an overview.
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_token.py:24]: I feel like this is not enough detail, see your ovirt example

https://github.com/ansible/ansible/blob/stable-2.9/lib/ansible/modules/cloud/ovirt/ovirt_auth.py#L19-L20

> which should be later used in all other oVirt/RHV modules, so all modules don't need to perform login and logout.

It is implied in its wording that it's intended that you use the module, and then don't pass the token value manually, because it goes through ansible_facts. You can say any variation on this, but it needs to be said.

File: awx_collection/plugins/modules/tower_token.py (hunk 4 above)
Comment (AlanCoding) [reply awx_collection/plugins/modules/tower_token.py]: doesn't look like that got pushed yet?

File: awx_collection/plugins/modules/tower_token.py (hunk 5 above)
Comment (AlanCoding) [reply awx_collection/plugins/modules/tower_token.py:98]: but then if the task "Create a new token using username/password" failed, then it would try to delete the 1st token twice, failing somewhat erroneously.

Understanding better what this is trying to do, I think it would be best to split it into 2 separate examples, where they both delete the token in the end with the `always:` block, because that matches better with what we're trying to recommend.

Also, when you do the `tower_job_list`, it isn't necessary to pass the `tower_oauthtoken` parameter, right? Or am I misunderstanding?

File: awx_collection/plugins/module_utils/tower_api.py [hunk 11]
```diff
@@ -487,6 +490,9 @@ def delete_if_needed(self, existing_item, on_delete=None):
                 item_name = existing_item['username']
             elif 'identifier' in existing_item:
                 item_name = existing_item['identifier']
+            elif item_type == 'o_auth2_access_token':
```
Comment (squidboylan) [inline awx_collection/plugins/module_utils/tower_api.py:521]: im not sure we'll hit this, testing locally when i output the token object i get:

```
TASK [tower_token : debug] ****************************************************************************************************************************************************************************************************************************************************************************
ok: [testhost] => {
    "msg": {                                                                                                                                                                                                                                                                                          
        "ansible_facts": {                                                                                                                                                                                                                                                                            
            "tower_token": {
                "application": null,                                                                                                                                                                                                                                                                  
                "created": "2020-06-01T18:48:22.790373Z",
                "description": "AWX-Collection-tests-tower_token-description-NtDlNNHIgxVWVYJp",                                                                                                                                                                                                       
                "expires": "3019-10-03T18:48:22.787511Z",
                "id": 461,                                                                                                                                                                                                                                                                            
                "modified": "2020-06-01T18:48:22.799678Z",                                                                                                                                                                                                                                            
                "refresh_token": null,
                "related": {                                                                                                                                                                                                                                                                          
                    "activity_stream": "/api/v2/tokens/461/activity_stream/",
                    "user": "/api/v2/users/1/"
                },
                "scope": "write",
                "summary_fields": {                                                                                                                                                                                                                                                                   
                    "user": {
                        "first_name": "",                                                                                                                                                                                                                                                             
                        "id": 1,
                        "last_name": "",                                                                                                                                                                                                                                                              
                        "username": "admin"
                    }
                },                                                                                                                                                                                                                                                                                    
                "token": "Zvk7PYb0eKusQn6fx0jTgGuvp8Yo9m",                                                                                                                                                                                                                                            
                "type": "o_auth2_access_token",                                                                                                                                                                                                                                                       
                "url": "/api/v2/tokens/461/",
                "user": 1
            }
        },
        "changed": true,
        "failed": false,
        "id": 461,
        "name": "unknown",                                                                                                                                                                                                                                                                            
        "token": "Zvk7PYb0eKusQn6fx0jTgGuvp8Yo9m",
        "warnings": [                                                                                                                                                                                                                                                                                 
            "You are running collection version devel but connecting to tower version 11.2.0"
        ]
    }                                                                                                                                                                                                                                                                                                 
}
```

and if it has a name we use the name, im having trouble locally deleting stuff, im not sure the delete actually works

File: awx_collection/tests/integration/targets/tower_token/tasks/main.yml (hunk 8 above)
Comment (squidboylan) [reply awx_collection/tests/integration/targets/tower_token/tasks/main.yml:66]: these tests still need to use `new_token` though right?

Comment (squidboylan) [reply awx_collection/tests/integration/targets/tower_token/tasks/main.yml:66]: oh nvm, looks like `tower_token` works, i didnt realize modules could just register variables but thinking about it now it makes sense

File: awx_collection/plugins/module_utils/tower_api.py (hunk 7 above)
Comment (squidboylan) [reply awx_collection/plugins/module_utils/tower_api.py:108]: i think that's suppsed to be `elif not 'token' in self.params.get('tower_oauthtoken'):` which gives us the old string behavior, my testing shows that the old string behavior is currently broken, im pretty sure that's why

Comment (john-westcott-iv) [reply awx_collection/plugins/module_utils/tower_api.py:108]: This has been updated, take a look and let me know what you think.

Comment (AlanCoding) [reply awx_collection/plugins/module_utils/tower_api.py:108]: 👍

File: awx_collection/plugins/modules/tower_token.py (hunk 9 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_token.py:94]: I'm a little torn on providing additional details about the return dict. The whole point of the module is to hide the login mechanics from someone. IMHO we should be trying to document "here is how you call this thing and pass it onto the next" vs. "here is how you call this thing, here are all of the internals of what you get back, now pass it onto the next". If there was other useful data within the dick maybe but I'm not sure the fields would provide any use. And people really should not have a need to "create" one of these dicts on their own to use for authentication (that would likely be more work than just cutting a token directly in Tower and using the string form). Do you see an advantage to documenting these files two fields?

As for the return 'token' value, I think that might just be an artifact from before getting the ansible_fact returned. Any objection to removing it from the response?

Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_token.py]: Technically this module is the CRUD interface to the tokens endpoint and no other module includes a RETURN statement. But I think this one should because it’s setting the ansible_fact for token usage which deviates form the norm. In addition, the token string is only available though this fact once and can't be obtained through the API on subsequent calls.

Because of that, I pushed a change which modified the module description to add:
```
    - In addition, the module sets an Ansible fact which can be passed into other
      tower_* modules as the parameter tower_oauthtoken. See examples for usage.
    - Because of the sensitive nature of tokens, the created token value is only available once
      through the Ansible fact. (See RETURN for details)
```

Then In the return we say:
```
tower_token:
  type: dict
  description: An Ansible Fact variable representing a Tower token object which can be used for auth in subsequent modules. See examples for usage.
  contains:
    token:
      description: The token that was generated. This token can never be accessed again, make sure this value is noted before it is lost.
      type: str
  returned: on successful create
```

However, I’m not sure we want to dive into the data structure (i.e. the ID). Technically, like all other modules, the ID is in the return. But people really shouldn't care what else is in the dict and they shouldn't be trying to create one on their own. 

I also removed the token in the response (that was just an artifact from before the ansible_fact was returned). 

Let me know what people think about these changes.

File: awx_collection/plugins/module_utils/tower_api.py (hunk 11 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/module_utils/tower_api.py:521]: I pushed some updated code for this. Let me know if this resolves the issue.

**Review by squidboylan — APPROVED** [review]
LGTM

File: awx_collection/plugins/modules/tower_token.py [hunk 12]
```diff
@@ -0,0 +1,168 @@
...
+'''
+
+RETURN = '''
+tower_token:
+  type: dict
+  description: An Ansible Fact variable representing a Tower token object which can be used for auth in subsequent modules. See examples for usage.
+  contains:
+    token:
+      description: The token that was generated. This token can never be accessed again, make sure this value is noted before it is lost.
+      type: str
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_token.py:113]: Again, I think documenting `id` (if that is what the key is) is a minimum requirement, because it is part of the function of the module `tower_token`, and a part of the interface between the module logic and the parameter structure passed into it.

File: awx_collection/plugins/modules/tower_token.py [hunk 13]
```diff
@@ -0,0 +1,168 @@
...
+    scope:
+      description:
+        - Allowed scopes, further restricts user's permissions. Must be a simple space-separated string with allowed scopes ['read', 'write'].
+      required: False
+      type: str
+      default: 'write'
+      choices: ["read", "write"]
+    existing_token:
+      description: An existing token (for use with state absent)
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_token.py]: People will absolutely mis-construe this. People will read "token" as the token value, but that doesn't match the fact that it's a dict.

This single field is expected to be a mirror of all of the module params combined into a dict. It's fine to say that, and much better than not saying it.

What about idempotency? Is this module expected to be idempotent, and if not, what is expected of the user in order to write a playbook that doesn't leave around stale tokens?

File: awx_collection/plugins/modules/tower_token.py [hunk 14]
```diff
@@ -0,0 +1,194 @@
...
+    description = module.params.get('description')
+    application = module.params.get('application')
+    scope = module.params.get('scope')
+    existing_token = module.params.get('existing_token')
+    existing_token_id = module.params.get('existing_token_id')
+    state = module.params.get('state')
+
+    if state == 'absent':
+        if not existing_token:
+            if not existing_token_id:
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_token.py]: it could make sense to mark these as mutually exclusive.

File: awx_collection/test/awx/test_token.py [hunk 15]
```diff
@@ -0,0 +1,28 @@
...
+        'tower_host': None,
+        'tower_username': None,
+        'tower_password': None,
+        'validate_certs': None,
+        'tower_oauthtoken': None,
+        'tower_config_file': None,
+    }
+
+    result = run_module('tower_token', module_args, admin_user)
+    assert result, result.get('changed')
```
Comment (AlanCoding) [inline awx_collection/test/awx/test_token.py]: I think is is backwards, `assert result` will always evaluate to true with a non-empty dict. You want to `assert result.get('changed'), result`, meaning that the result will be printed if it's not changed.

File: awx_collection/test/awx/test_token.py [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+
+    tokens = OAuth2AccessToken.objects.filter(description='barfoo')
+    assert len(tokens) == 1, tokens[0].description == 'barfoo'
```
Comment (AlanCoding) [inline awx_collection/test/awx/test_token.py]: maybe assert that the token's scope is read?

File: awx_collection/plugins/modules/tower_token.py (hunk 14 above)
Comment (john-westcott-iv) [reply awx_collection/plugins/modules/tower_token.py]: Done, also added required_if so that one of them must be present if state == 'absent'

File: awx_collection/test/awx/test_token.py (hunk 15 above)
Comment (john-westcott-iv) [reply awx_collection/test/awx/test_token.py]: Thanks, fixed.

File: awx_collection/test/awx/test_token.py (hunk 16 above)
Comment (john-westcott-iv) [reply awx_collection/test/awx/test_token.py]: Thanks, resolved.

File: awx_collection/tests/integration/targets/tower_token/tasks/main.yml [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+- assert:
+    that:
+      - results is failed
+      - '"The provided tower_oauthtoken type was not valid (list). Valid options are str or dict." == results.msg'
+
+- name: Try to delete a token with no existing_token or existing_token_id
+  tower_token:
+    state: absent
+  register: results
+  ignore_errors: True
```
Comment (AlanCoding) [inline awx_collection/tests/integration/targets/tower_token/tasks/main.yml]: I think this is the last issue from CI

```
2020-06-08 10:53:56.769399 | static | linters run-test: commands[1] | yamllint -s .
2020-06-08 10:53:59.792596 | static | ./awx_collection/tests/integration/targets/tower_token/tasks/main.yml
2020-06-08 10:53:59.792784 | static |   34:18     warning  truthy value should be one of [false, true]  (truthy)
2020-06-08 10:53:59.792890 | static |   48:18     warning  truthy value should be one of [false, true]  (truthy)
```

Comment (john-westcott-iv) [reply awx_collection/tests/integration/targets/tower_token/tasks/main.yml]: This one gets me all the time.

File: awx_collection/test/awx/test_token.py [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+        'validate_certs': None,
+        'tower_oauthtoken': None,
+        'tower_config_file': None,
+    }
+
+    result = run_module('tower_token', module_args, admin_user)
+    assert result.get('changed'), result
+
+    tokens = OAuth2AccessToken.objects.filter(description='barfoo')
+    assert len(tokens) == 1, 'Rokens with description of barfoo != 0: {0}'.format(len(tokens))
```
Comment (AlanCoding) [inline awx_collection/test/awx/test_token.py]: Rokens?

File: awx_collection/plugins/modules/tower_token.py [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+
+  always:
+    - name: Delete our Token with the token we created
+      tower_token:
+        existing_token: "{{ tower_token }}"
+        state: absent
+      when: tower_token is defined
+
+- name: Delete a token by its id
+  tower_toekn:
```
Comment (AlanCoding) [inline awx_collection/plugins/modules/tower_token.py]: `tower_toekn` -> `tower_token`

File: awx_collection/plugins/modules/tower_token.py (hunk 9 above)
Comment (AlanCoding) [reply awx_collection/plugins/modules/tower_token.py]: ok, resolved

this was the most important things out of the various discussions here

**Review by AlanCoding — APPROVED** [review]
